Mathlib/Tactic/Basic.lean

```diff
@@ -188,10 +188,49 @@ macro_rules
 
 macro (name := «sorry») "sorry" : tactic => `(exact sorry)
 
--- TODO `n:num` should be an optional argument; if missing, repeat until failure
-elab "iterate " n:num seq:tacticSeq : tactic => do
-  for i in [:n.toNat] do
-    evalTactic seq
+/--
+`iterate_at_most n t` runs the tactic `t` at most `n` times, or until failure,
+whichever comes first.
+Always succeeds, and returns at list of the succesful results.
+-/
+def iterate_at_most : Nat → TacticM α → TacticM (List α)
+| 0,       t => []
+| (n + 1), t => do
+  try
+    (←t) :: (←iterate_at_most n t)
+  catch _ =>
+    []
+
+/--
+`iterate_at_most' n t` runs the tactic `t` at most `n` times, or until failure,
+whichever comes first, and always succeeds.
+-/
+def iterate_at_most' (n : Nat) (t : TacticM α) : TacticM Unit := do
+  _ ← iterate_at_most n t
+
+/--
+`iterate_exactly n t` runs the tactic `t` at exactly `n` times.
+Fails if any iteration fails, and otherwise returns a list of the `n` successful results.
+-/
+def iterate_exactly : Nat → TacticM α → TacticM (List α)
+| 0, t     => []
+| (n+1), t => do (←t) :: (←iterate_exactly n t)
+
+/--
+`iterate_exactly n t` runs the tactic `t` at exactly `n` times.
+Fails if any iteration fails.
+-/
+def iterate_exactly' (n : Nat) (t : TacticM α) : TacticM Unit := do
+  _ ← iterate_exactly n t
+
+/--
+`iterate n { ... }` runs the tactic block exactly `n` times.
+`iterate { ... }` runs the tactic block repeatedly until failure.
+-/
+elab "iterate " n:(num)? ppSpace seq:tacticSeq : tactic => do
+  match n.getOptional? with
+  | none => iterate_at_most' 10000 (evalTactic seq)
+  | some n => iterate_exactly' n.toNat (evalTactic seq)
```

Comment:
I'm not sure it's a good idea to copy the Lean 3 implementation (at the very least we should get rid of the old naming convention).  The concise Lean 4 way (TM) to do this seems to be this:
```lean
syntax "iterate " (num)? ppSpace tacticSeq : tactic
macro_rules
  | `(tactic|iterate $seq:tacticSeq) =>
    `(tactic|try ($seq:tacticSeq); iterate $seq:tacticSeq)
  | `(tactic|iterate $n $seq:tacticSeq) =>
    match n.toNat with
    | 0 => `(tactic| skip)
    | n+1 => `(tactic|($seq:tacticSeq); iterate $(quote n) $seq:tacticSeq)
```

Reply:
What about the versions where we want to collect the results as a `List α`? I don't know how to implement those in this manner. :-(

Certainly we want to have the "plumbing" versions of tactics available too -- without them we very quickly find ourselves unable to implement the next layer of tactics. (I wouldn't die on the hill of these particular tactics, I guess) And then surely the porcelain versions should use the plumbing!

Reply:
I'm not sold on the Lean 4 way (TM) of using syntax macros for everything. It is too weakly typed for my taste, and I think it will be less efficient than a regular "plumbing" tactic. It's cute that it can be done, and it's great for one-off tactics, but there is no way I'm implementing `normNum` as a syntax macro. I think we should just port the `tactic.core` tactics to regular functions in `MetaM`, and most interactive tactics should be implemented as an `elab` which calls some plumbing tactic.

Reply:
Okay, how about the new version, where I have:

1. switched to using `elab_rules` with two cases, rather than `match n.getOptional? with`
2. added a note:
* explaining that having the plumbing tactics is important
* and preserving @gebner's suggestion as an example (of which we desperately need more still!)

Reply:
> It's cute that it can be done, and it's great for one-off tactics, but there is no way I'm implementing normNum as a syntax macro.

I 90% agree with `normNum` (i.e., `normNum1` should be implemented using elab), but I also think that `iterate` is a one-off tactic.

Practically speaking, there's little difference between implementing `iterate` as a macro or as a plumbing tactic (besides being 10x shorter).  (If you're worried about the performance of macro expansion here, I'd like you to think about what `iterate refine ⟨1, ?_⟩` does.)

One potential (future) advantage of macros is that we might be able to hook up macro expansion as a code action in the editor so that you can replace iterate by its expansion.  This could be nice particularly for high-level and control tactics.

Reply:
@gebner, what about the use case where someone wants to collect the list of results of an iterate? Perhaps this isn't a really essential use case, but it's certainly something I can imagine wanting, and it seems implementing `iterate` as a macro doesn't allow for this.

Reply:
That seems like a theoretical question.  I just looked through mathlib and I could only find a *single* non-interactive use of `iterate` (namely in `intros1`).

In Lean 4 the `iterate` function is also less ergonomic than in Lean 3 because it doesn't play nice with do-notation.  For example you can't assign variables inside iterate like this:
```
while true do
  try
    let (n, mvar') <- intro1 mvar
    mvar := mvar'
    ns := ns.push n
  catch _ =>
    break
```

Reply:
Okay! I will update to just use the macro versions.